counters.hh

```diff
@@ -328,17 +327,14 @@ private:
             return it;
         }
         bool operator==(const shard_iterator& other) const noexcept {
-            return _current == other._current;
-        }
-        bool operator!=(const shard_iterator& other) const noexcept {
-            return !(*this == other);
+            return _pos == other._pos;
         }
     };
 public:
     boost::iterator_range<shard_iterator> shards() const {
         auto value = _cell.value();
-        auto begin = shard_iterator(value);
-        auto end = shard_iterator();
+        auto begin = shard_iterator(value, 0);
+        auto end = shard_iterator(value, value.size());
```

Comment:
This looks like something that the iterator can do inside the constructor (setting pos).